### PR TITLE
Implementation of unit conversion in the CV module 

### PR DESCRIPTION
**Pull request by DunklesArchipel**
Addresses several issues:

- [x] #41
- [x] #42 
- [x] #43 
- [x] #44 
- [x] #45  
    - [x] conversion to csv 
    - [x] create also an output json
- [x] #46 can be done after #41
- [x] #47 related to #44 : The columns in the df are all in SI units
- [x] Add tests for output csv
- [x] Add test for output json

### CODE REVIEW

**Comment by saraedum** [comment]
@DunklesArchipel you might want to pull in the latest changes from my transform branch (`git pull saraedum transform`) to make the tests pass again. Note that `get_labels` has been renamed to `get_texts` as discussed.

**Comment by DunklesArchipel** [comment]
This is a good starting point and the program does what it should. 

Some aspects should maybe later be solved differently. For example the functions that create the `U`, `I`/`j` or `t` axis could be classmethods. It is not uncommon that I is plotted vs U and hence not only relevant for CVs. The functions could be refused for other modules.

File: svgdigitizer/__main__.py [hunk 1]
```diff
@@ -31,19 +31,46 @@ def cli(): pass
...
+        cv = CV(SVGPlot(SVG(open(svg, 'rb')), sampling_interval=sampling_interval), metadata=metadata)
+
+    else:
+        cv = CV(SVGPlot(SVG(open(svg, 'rb')), sampling_interval=sampling_interval))
+
+    from pathlib import Path
+    cv.cv_df.to_csv(Path(svg).with_suffix('.csv'), index=False)
+    import json
+    with open(Path(svg).with_suffix('.json'), "w") as outfile:
+        json.dump(cv.metadata_out, outfile)
```
Comment (saraedum) [inline svgdigitizer/__main__.py:71]:
```suggestion
@click.option('--metadata', type=click.File("rb"), default=None, help='yaml file with metadata')
@click.argument('svg', type=click.Path(exists=True))
def cv(svg, sampling_interval, metadata):
    import yaml
    from svgdigitizer.svgplot import SVGPlot
    from svgdigitizer.svg import SVG
    from svgdigitizer.electrochemistry.cv import CV
    if metadata:
        metadata = yaml.load(metadata, Loader=yaml.FullLoader)

    cv = CV(SVGPlot(SVG(open(svg, 'rb')), sampling_interval=sampling_interval), metadata=metadata)

    from pathlib import Path
    cv.cv_df.to_csv(Path(svg).with_suffix('.csv'), index=False)

    import json
    with open(Path(svg).with_suffix('.json'), "w") as outfile:
        json.dump(cv.metadata_out, outfile)
```

I think this can be simplified like this.

File: svgdigitizer/electrochemistry/cv.py [hunk 2]
```diff
@@ -18,111 +18,166 @@
...
+unit_typos = {'uA / cm2': ['uA / cm2', 'uA / cm²', 'µA / cm²', 'µA cm⁻²', 'uA cm-2', 'uA / cm2'],
+              'A / cm2': ['A / cm2', 'A cm⁻²', 'A cm-2', 'A / cm2'],
+              'A': ['A', 'ampere', 'amps', 'amp'],
+              'mV': ['milliV', 'millivolt', 'milivolt', 'miliv', 'mV'],
+              'V': ['V', 'v', 'Volt', 'volt'],
+              'V / s': ['V s-1', 'V / s'],
+              'mV / s': ['mV / s', 'mV s-1', 'mV/s']}
 
 
 class CV():
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:45]: Can we add some documentation here and a usage example?

File: svgdigitizer/electrochemistry/cv.py [hunk 3]
```diff
@@ -18,111 +18,166 @@
...
-    def __init__(self, metadata, svgplot):
+    def __init__(self, svgplot, metadata=None):
         """
         metadata: dict
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:48]: That's a lie now because `metadata` can be `None` apparently. It's best to delete this comment altogether.

File: svgdigitizer/electrochemistry/cv.py [hunk 4]
```diff
@@ -18,111 +18,166 @@
...
+        return {'x': {'dimension': 'U',
+                      'unit': 'V'},
+                'y': {'dimension': 'j' if 'm2' in str(self.get_axis_unit('y')) else 'I',
+                      'unit': 'A / m2' if 'm2' in str(self.get_axis_unit('y')) else 'A'}}
 
-        # TODO: should be j if both current and normalized to are given in the yaml file
-        self.ylabel = 'I'
+    def get_axis_unit(self, axis):
+        r'''
+        Replaces the units derived from the svg file into astropy units.
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:63]: This comment is not that helpful for somebody who looks at the code for the first time imho. `Replaces` indicates that this actually changes the state of the `CV` somehow. But it does not.
So what does this method do? It returns the unit of the `axis` axis in a certain normalized format. So it should probably say `Return the unit of the axis `axis` ...`.

File: svgdigitizer/electrochemistry/cv.py [hunk 5]
```diff
@@ -18,111 +18,166 @@
...
+        '''
+        unit = self.svgplot.units[axis]
 
-        # TODO: All the rest in the init is presumably not necessary
-        self.description = self.metadata['figure description']
+        return self.get_correct_unit(unit)
 
-        self.xunit = self.description['potential scale']['unit']
+    @classmethod
+    def get_correct_unit(cls, unit):
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:70]: Is there any reason why the `typos` above are not inlined in this method? Then the comment above could probably also be integrated into a docstring for this method. Probably, without explaining how the implementation works exactly, but rather explaining what this method actually does.

File: svgdigitizer/electrochemistry/cv.py [hunk 6]
```diff
@@ -18,111 +18,166 @@
...
 
-        self.yunit = self.description['current']['unit']
+        raise ValueError(f'Unknown Unit {unit} on Axis {axis}')
 
-        self.get_rate()
+    @property
+    @cache
+    def rate(self):
+        r'''
+        Return a rate based on a label in the SVG file.
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:81]: Could you elaborate in the docstring what a `rate` is?

It's also usually good to split such comments into "what does it do" and "how does it do it" such as:
```
Return the scan rate of the plot.

The scan rate is read from a `<text>` in the SVG file such as `<text>scan rate: 1abc</text>`.
```

File: svgdigitizer/electrochemistry/cv.py [hunk 7]
```diff
@@ -18,111 +18,166 @@
...
+        '''
+        rates = self.svgplot.svg.get_texts('(?:scan rate|rate): (?P<value>-?[0-9.]+) *(?P<unit>.*)')
+        # To Do
+        # assert that only one label contains the scan rate
+        # asstert that a rate is available at all
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:86]:
```suggestion
        # TODO: assert that only one label contains the scan rate (see issue #ABC)
        # TODO: assert that a rate is available at all (see issue #ABC).
```

File: svgdigitizer/electrochemistry/cv.py [hunk 8]
```diff
@@ -18,111 +18,166 @@
...
-        Return rate based on the x coordinate units.
+        return rate
 
-        At the moment we simply use the value.
-        '''
-        self.rate = self.description['scan rate']['value']
-        return self.rate
+    @property
+    @cache
+    def cv_df(self):
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:97]: Please add a docstring here. It is not clear what `cv_df` is otherwise.

File: svgdigitizer/electrochemistry/cv.py [hunk 9]
```diff
@@ -18,111 +18,166 @@
...
-        # Create potential columns
-        self.df = self.create_df_U_axis(self.svgplot.dfs[0])
-        # Create current columns
-        self.df = pd.concat(
-            [self.df, self.create_df_I_axis(self.svgplot.dfs[0])],
-            axis=1)
+        # Create time axis.
+        df['t'] = self.create_df_time_axis(df)
+        # Rearrange columns.
+        df = df[['t', 'U', self.axis_properties['y']['dimension']]]
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:107]: These `create_` calls are a bit strange. They return a dataframe but at the same time they mutate their argument. Would it maybe be easier to only mutate the argument? I.e., say `self._add_U_axis(self.svgplot.df)` here instead?

Comment (saraedum) [reply svgdigitizer/electrochemistry/cv.py:107]: Btw., it would be nice to prefix internal methods with `_`. This way they do not show up in tab completion when using Jupyter and such. Usually, no one wants to call `create_df_U_axis` directly.

File: svgdigitizer/electrochemistry/cv.py [hunk 10]
```diff
@@ -18,111 +18,166 @@
...
 
-        # create time axis
-        self.df['t'] = self.create_df_time_axis(self.svgplot.dfs[0])
+        return df
 
     def create_df_U_axis(self, df):
         r'''
         Create voltage axis in the dataframe based on the units given in the
         figure description.
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:114]: In all of these docstrings, that this comes from the figure description is not really true anymore. It's maybe better to not be explicit about this and refer the reader to the correct place, i.e., just say: ``Add voltage column to the dataframe `df`, based on the :meth:`get_axis_unit`.``

File: svgdigitizer/electrochemistry/cv.py [hunk 11]
```diff
@@ -18,111 +18,166 @@
...
-            df_['I_A'] = df['I']
 
-        if self.yunit == 'mA':
-            df_['I_A'] = df['I']/1E3
+        df_ = df.copy()
+        q = 1 * self.get_axis_unit('y')
 
-        if self.yunit == 'uA':
-            df_['I_A'] = df['I']/1E6
+        # Verify if the y data is current ('A') or current density ('A / cm2')
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:132]:
```suggestion
        # Distinguish whether the y data is current ('A') or current density ('A / cm2')
```

File: svgdigitizer/electrochemistry/cv.py (hunk 8 above)
Comment (saraedum) [reply svgdigitizer/electrochemistry/cv.py:97]: Maybe just `dataframe(self)` or `df(self)` instead?

File: svgdigitizer/electrochemistry/cv.py [hunk 12]
```diff
@@ -18,111 +18,166 @@
...
-        df_['t'] = df_['cumdeltaU']/self.get_rate()
-        return df_['t']
+        df = df.copy()
+        df['deltaU'] = abs(df['U'].diff().fillna(0))
+        df['cumdeltaU'] = df['deltaU'].cumsum()
+        df['t'] = df['cumdeltaU']/float(self.rate.to(u.V / u.s).value)
+
+        return df[['t']]
 
     def plot_cv(self):
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:154]:
```suggestion
    def plot(self):
```

File: svgdigitizer/electrochemistry/cv.py [hunk 13]
```diff
@@ -18,111 +18,166 @@
...
-        self.df.plot(x='U_V', y='I_uA')
-        plt.xlabel(f'{self.xlabel} / {self.xunit}')
-        plt.ylabel(f'{self.ylabel} / {self.yunit}')
+        self.cv_df.plot(x=self.axis_properties['x']['dimension'], y=self.axis_properties['y']['dimension'])
+        plt.axhline(linewidth=1, linestyle=':', alpha=0.5)
+        plt.xlabel(self.axis_properties['x']['dimension'] + ' / ' + str(self.axis_properties['x']['unit']))
+        plt.ylabel(self.axis_properties['y']['dimension'] + ' / ' + str(self.axis_properties['y']['unit']))
+
+    @property
+    def metadata_out(self, comment=''):
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:161]: Why `out`? Why not just `metadata`? If you want to avoid the clash with the user supplied metadata, just call that one `_metadata` since the user is, e.g., in a Jupyter session, not supposed to interact with that metadata dict anymore anyway.

File: svgdigitizer/electrochemistry/cv.py [hunk 14]
```diff
@@ -18,111 +18,166 @@
...
+        # Add description
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:162]:
```suggestion
```

File: svgdigitizer/electrochemistry/cv.py [hunk 15]
```diff
@@ -18,111 +18,166 @@
...
+        assert type(comment) == str, 'Comment must be a string'
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:163]:
```suggestion
```
We don't usually do type checks in Python. All that matters is that what the user provided behaves like a string. So when you use it, just say `= str(comment)`.

File: svgdigitizer/electrochemistry/cv.py [hunk 16]
```diff
@@ -18,111 +18,166 @@
...
+        metadata = self.metadata
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:164]: This does not create a copy but will mutate `self.metadata`.

```suggestion
        metadata = self.metadata.copy()
```

File: svgdigitizer/electrochemistry/cv.py [hunk 17]
```diff
@@ -18,111 +18,166 @@
...
+        if 'figure description' not in self.metadata:
+            metadata['figure description'] = {}
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:166]:
```suggestion
        metadata.setdefault('figure description', {})
```

File: svgdigitizer/electrochemistry/cv.py [hunk 18]
```diff
@@ -18,111 +18,166 @@
...
+
+        metadata['figure description']['type'] = 'digitized'
+        metadata['figure description']['measurement type'] = 'CV'
+        metadata['figure description']['scan rate'] = {'value': self.rate.value, 'unit': str(self.rate.unit)}
+        if 'potential scale' not in metadata['figure description']:
+            metadata['figure description']['potential scale'] = {}
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:172]: Use `setdefault` instead.

File: svgdigitizer/electrochemistry/cv.py [hunk 19]
```diff
@@ -18,111 +18,166 @@
...
+        metadata['figure description']['potential scale']['unit'] = str(self.get_axis_unit('x'))
+        # Implement: Get the reference from the text labels of the axis
+        # metadata['figure description']['potential scale']['reference'] = get_from_ax_labels
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:175]: Is there an issue for this? We should write `TODO: ... #ABC` so it is clear that something is missing here.
There shuold be no commented-out code. It's afterwards almost impossible to understand why it was commented out ;)

File: svgdigitizer/electrochemistry/cv.py [hunk 20]
```diff
@@ -18,111 +18,166 @@
...
+        metadata['figure description']['current'] = {'unit': str(self.get_axis_unit('y'))}
+        metadata['figure description']['comment'] = comment
+
+        return metadata
 
     def create_csv(self, filename):
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:181]: This method can be deleted I think. We are not using it anymore.

**Review by saraedum — CHANGES_REQUESTED** [review]
I made a quick pass over your changes. Mostly there are some style issues that can be easily improved I hope.
Let me finish #53. Without it merged, it's a bit hard to review this as my changes are mixed in here.

File: svgdigitizer/__main__.py [hunk 21]
```diff
@@ -31,19 +31,46 @@ def cli(): pass
...
     from svgdigitizer.svgplot import SVGPlot
-    plot = SVGPlot(open(svg, 'rb'), sampling_interval=sampling_interval)
+    from svgdigitizer.svg import SVG
+    plot = SVGPlot(SVG(open(svg, 'rb')), sampling_interval=sampling_interval)
     from pathlib import Path
     plot.df.to_csv(Path(svg).with_suffix('.csv'), index=False)
 
 
+@click.command()
+@click.option('--sampling_interval', type=float, default=None, help='specify sampling interval (for now in mV)')
```
[the remark this answers is not in the document]
Comment (DunklesArchipel) [reply svgdigitizer/__main__.py:50]: sure!

File: svgdigitizer/__main__.py (hunk 1 above)
Comment (DunklesArchipel) [reply svgdigitizer/__main__.py:71]: I was not sure what `click.type(..)` would do. :)
The rest is fine.

File: svgdigitizer/electrochemistry/cv.py (hunk 3 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:48]: done

File: svgdigitizer/electrochemistry/cv.py (hunk 4 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:63]: I removed this function since as you mention it does nothing else than assigning `self.svgplot.units['x']` to `unit`.
Instead I use `CV.get_correct_unit(self.svgplot.units['x'])`. I guess that makes more sense.

File: svgdigitizer/electrochemistry/cv.py (hunk 5 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:70]: I wrote a doctest. I guess it contains to much information. For me it would be helpful. :)

File: svgdigitizer/electrochemistry/cv.py [hunk 22]
```diff
@@ -18,111 +18,199 @@
...
-        self.metadata = metadata
-        self.svgplot.create_df()
-
-        # TODO: These labels should either be extracted from the svg or yaml file
-        self.xlabel = 'U'
-
-        # TODO: should be j if both current and normalized to are given in the yaml file
-        self.ylabel = 'I'
+        rates = self.svgplot.svg.get_texts('(?:scan rate|rate): (?P<value>-?[0-9.]+) *(?P<unit>.*)')
+        # To Do: assert that only one label contains the scan rate (see issue #X)
```
Comment (DunklesArchipel) [inline svgdigitizer/electrochemistry/cv.py:119]: Partially revised up to this line.

File: svgdigitizer/electrochemistry/cv.py (hunk 8 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:130]: I hope my explanation is sufficient

File: svgdigitizer/electrochemistry/cv.py (hunk 17 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:199]: excellent I tried to find something exactly like this. :)

File: svgdigitizer/electrochemistry/cv.py [hunk 23]
```diff
@@ -19,110 +19,199 @@
...
+                'y': {'dimension': 'j' if 'm2' in str(CV.get_axis_unit(self.svgplot.units['x'])) else 'I',
+                      'unit': 'A / m2' if 'm2' in str(CV.get_axis_unit(self.svgplot.units['y'])) else 'A'}}
 
-        # TODO: should be j if both current and normalized to are given in the yaml file
-        self.ylabel = 'I'
+    @classmethod
+    def get_axis_unit(cls, unit):
+        r"""
+        Convert a string containing a unit into a string that can be used by the astropy.unit module.
+        For example 'uA cm-2' should become 'uA / cm2'
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:47]: That's not true actually. It does not return a string but a unit.
```suggestion
        Return `unit` as an astropy unit.
        
        This method normalizes unit names, e.g., it rewrites 'uA cm-2' to 'uA / cm2' which astropy understands.
```

File: svgdigitizer/electrochemistry/cv.py [hunk 24]
```diff
@@ -19,110 +19,199 @@
...
+        'A / cm2'
 
-        self.modify_df()
+        These string can directly be converted to a unit
 
-    def get_rate(self):  # TODO: probably not required
-        r'''
-        Return rate based on the x coordinate units.
+        >>> u.Unit('uA / cm2')
+        Unit("uA / cm2")
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:66]: This seems to be unrelated to this method. I would remove this section.

File: svgdigitizer/electrochemistry/cv.py [hunk 25]
```diff
@@ -19,110 +19,199 @@
...
-        # Create potential columns
-        self.df = self.create_df_U_axis(self.svgplot.dfs[0])
-        # Create current columns
-        self.df = pd.concat(
-            [self.df, self.create_df_I_axis(self.svgplot.dfs[0])],
-            axis=1)
+        """
+        rates = self.svgplot.svg.get_texts('(?:scan rate|rate): (?P<value>-?[0-9.]+) *(?P<unit>.*)')
+        # To Do: assert that only one label contains the scan rate (see issue #58)
+        # To Do: assert that a rate is available at all (see issue #58)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:125]:
```suggestion
        # TODO: assert that only one label contains the scan rate (see issue #58)
        # TODO: assert that a rate is available at all (see issue #58)
```
Can we always write `TODO`? I think that this is the format that most editors highlight specially.

File: svgdigitizer/electrochemistry/cv.py [hunk 26]
```diff
@@ -19,110 +19,199 @@
...
+        The time axis can only be created when a (scan) rate is given in the plot, i.e., 50 mV /s.
+        """
+        # Add a potential axis.
+        df = self._add_U_axis(self.svgplot.df)
 
-        return df_[['U_V', 'U_mV']]
+        df = self._add_I_axis(self.svgplot.df)
 
-    def create_df_I_axis(self, df):
+        df = self._add_time_axis(df)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:150]: That's really odd. Do these methods modify the argument or don't they? If they do, then there's no need to write `df = `.  If they don't the second line should operate on `df` and not on `self.svgplot.df`.

File: svgdigitizer/electrochemistry/cv.py [hunk 27]
```diff
@@ -19,110 +19,199 @@
...
+
+        # Rearrange columns.
+        df = df[['t', 'U', self.axis_properties['y']['dimension']]]
+
+        return df
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:155]:
```suggestion
        return df[['t', 'U', self.axis_properties['y']['dimension']]]
```

File: svgdigitizer/electrochemistry/cv.py [hunk 28]
```diff
@@ -19,110 +19,199 @@
...
+        # to convert the potential values in the df to V
+        df['U'] = df['x'] * q.to(u.V).value
+
+        return df
 
-        if self.yunit == 'mA':
-            df_['I_A'] = df['I']/1E3
+    def _add_I_axis(self, df):
+        r'''
+        Add current or current desnity column to the dataframe `df`, based on the :meth:`get_axis_unit`.
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:170]:
```suggestion
        Add current or current density column to the dataframe `df`, based on the :meth:`get_axis_unit` of the y axis.
```

File: svgdigitizer/electrochemistry/cv.py [hunk 29]
```diff
@@ -19,110 +19,199 @@
...
+        # Rearrange columns.
+        df = df[['t', 'U', self.axis_properties['y']['dimension']]]
+
+        return df
+
+    def _add_U_axis(self, df):
         r'''
-        Create current or current density axis in the dataframe based on the
-        units given in the figure description.
+        Add voltage column to the dataframe `df`, based on the :meth:`get_axis_unit`.
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:159]:
```suggestion
        Add voltage column to the dataframe `df`, based on the :meth:`get_axis_unit` of the x axis.
```

File: svgdigitizer/electrochemistry/cv.py [hunk 30]
```diff
@@ -19,110 +19,199 @@
...
         '''
-        df_ = df.copy()
-        if self.yunit == 'A':
-            df_['I_A'] = df['I']
+        q = 1 * CV.get_axis_unit(self.svgplot.units['x'])
+        # Convert the axis unit to SI unit V and use the value
+        # to convert the potential values in the df to V
+        df['U'] = df['x'] * q.to(u.V).value
+
+        return df
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:166]:
```suggestion

        return df
```
I would just make all these methods mutation. Currently, that's inconsistent.

Comment (saraedum) [reply svgdigitizer/electrochemistry/cv.py:166]: Same for the other `_add` methods.

File: svgdigitizer/electrochemistry/cv.py [hunk 31]
```diff
@@ -19,110 +19,199 @@
...
+        metadata.setdefault('figure description', {})
+        metadata['figure description']['type'] = 'digitized'
+        metadata['figure description']['measurement type'] = 'CV'
+        metadata['figure description']['scan rate'] = {'value': self.rate.value, 'unit': str(self.rate.unit)}
+        metadata['figure description'].setdefault('potential scale', {})
+        metadata['figure description']['potential scale']['unit'] = str(CV.get_axis_unit(self.svgplot.units['x']))
+        # TODO: Get the reference from the text labels of the axis #59
+        # such as: metadata['figure description']['potential scale']['reference'] = self.axis_properties['x']['reference']
+        metadata['figure description']['current'] = {'unit': str(CV.get_axis_unit(self.svgplot.units['y']))}
+        metadata['figure description']['comment'] = str(comment)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:215]:
```suggestion
        if comment: metadata['figure description']['comment'] = str(comment)
```
Do we want a "comment" even if there's none?

File: test/data/Author_YYYY_FirstTitleWord_Page_fignr_identifier.yaml [hunk 32]
```diff
@@ -1,6 +1,5 @@
-# This file can exist with the "source" information only, the entry data exists should be marked as "False" 
-# to indcate that neither the metadata nor the actual data were digitized or provided.
-# use "null" to idicate that a value is not available in the literature
+# This file can exist with the "source" information only
+# Figure description is adapted based on SVG input
```
Comment (saraedum) [inline test/data/Author_YYYY_FirstTitleWord_Page_fignr_identifier.yaml:2]: I don't understand this intro anymore. Should we delete it from these samples?

File: test/data/xy_rate.yaml [hunk 33]
```diff
@@ -0,0 +1,103 @@
+# This file can exist with the "source" information only
+# 'Figure description' is adapted based on SVG input
```
Comment (saraedum) [inline test/data/xy_rate.yaml:2]: I would delete this intro. I don't understand it at least.

File: test/data/xy_rate.yaml [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+source: 
+    doi: https://doi.org/10.1016/0039-6028(85)90985-9
+    bib: Beden_1985_Voltammetric_833 # AuthorName_YYYY_FirstWordTitle_pageNr. No prepositions. Fingers crossed that there is no duplicate.
+    figure: 2b # depending on the subfigure labels use 1, 1b, or 1_3
+    curve: First cycle # use a unique short description to identify the curve in the figure
+    
+curator:
+    name: Albert Engstfeld
+    orcid: https://orcid.org/0000-0002-9686-3948
+    digitized: 2021.06.01 # YYYY-MM-DD
```
Comment (saraedum) [inline test/data/xy_rate.yaml:14]: The value does not match the comment.

File: test/data/xy_rate.yaml [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+                  value: 0.1
+                  unit: M #[M, mM, µM, g kg-1, ...] 
+        ph: 
+            value: 13
+            uncertainty: 0.1
+        temperature: # provide the temperature as it is given in the article. Assume room temperature if nothing is provided
+            value: 298.15 
+            unit: K # [K, C, F]
+    electrodes:
+        configuration: three # number of electrodes
```
Comment (saraedum) [inline test/data/xy_rate.yaml:77]: Are we sure we want to spell out numbers here?

File: test/data/xy_rate.yaml [hunk 36]
```diff
@@ -0,0 +1,103 @@
...
+            shape: # optional
+                type: cylinder # bead/sphere
+                height:
+                    value: 4
+                    unit: mm
+                diameter:
+                    value: 4
+                    unit: mm
+            geometric electrolyte contact area: 
+                value: null
```
Comment (saraedum) [inline test/data/xy_rate.yaml:101]: Do we want to support `null + unit`? Or should this say `0 + unit`?

**Review by saraedum — CHANGES_REQUESTED** [review]

**Comment by saraedum** [comment]
Apart from some very minor changes, I think only the `_add` confusion needs to be cleaned up here.

File: svgdigitizer/electrochemistry/cv.py (hunk 31 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:214]: It could probably be omitted. I also think that this field should be added to the yaml file, when the svg is created. In that case, it should not even appear in this `metadata`..
The problem is related to the standard yaml file below, which should probably be fixed separately. 
For the time being, I suggest removing `comment` as an argument and also remove this line.

File: test/data/Author_YYYY_FirstTitleWord_Page_fignr_identifier.yaml (hunk 32 above)
Comment (DunklesArchipel) [reply test/data/Author_YYYY_FirstTitleWord_Page_fignr_identifier.yaml:2]: Can we discuss this file separately, since it has no consequences on this pull request?
The file structure and content really depend on the workflow. This was a second draft we created a while ago. 

Among other issues, I also don't understand why it is located in the test folder.

File: test/data/xy_rate.yaml (hunk 36 above)
Comment (DunklesArchipel) [reply test/data/xy_rate.yaml:101]: We thought `null` was a good option when the value was unknown.

File: test/data/xy_rate.yaml (hunk 35 above)
Comment (DunklesArchipel) [reply test/data/xy_rate.yaml:77]: Not sure if we had anything in mind. 3 should be equally fine.

File: test/data/xy_rate.yaml (hunk 34 above)
Comment (DunklesArchipel) [reply test/data/xy_rate.yaml:14]: This is related to issue #56

File: svgdigitizer/electrochemistry/cv.py (hunk 24 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:67]: Sure.

File: svgdigitizer/electrochemistry/cv.py (hunk 2 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py:29]: See issue #60

File: svgdigitizer/electrochemistry/cv.py (hunk 26 above)
Comment (DunklesArchipel) [reply svgdigitizer/electrochemistry/cv.py]: I hope I did it right. With the new approach `self.svgplot.df` should not be modified.

**Comment by DunklesArchipel** [comment]
> Apart from some very minor changes, I think only the `_add` confusion needs to be cleaned up here.

I included your changes. The part on the standard yaml file should be discussed separately. Currently that file is not used anywhere in the `svgdigitizer`.

File: svgdigitizer/electrochemistry/cv.py [hunk 37]
```diff
@@ -19,110 +19,183 @@
...
-        csvfile = Path(filename).with_suffix('.csv')
-        self.df.to_csv(csvfile, index=False)
+        df['deltaU'] = abs(df['U'].diff().fillna(0))
+        df['cumdeltaU'] = df['deltaU'].cumsum()
+        df['t'] = df['cumdeltaU'] / float(self.rate.to(u.V / u.s).value)
+
+        return df
+
+    def plot(self):
+        self.cv_df.plot(x=self.axis_properties['x']['dimension'], y=self.axis_properties['y']['dimension'])
```
Comment (DunklesArchipel) [inline svgdigitizer/electrochemistry/cv.py:182]:
```suggestion
        self.df.plot(x=self.axis_properties['x']['dimension'], y=self.axis_properties['y']['dimension'])
```

File: svgdigitizer/electrochemistry/cv.py [hunk 38]
```diff
@@ -19,110 +19,183 @@
...
-        figure description.
-        '''
-        df_ = df.copy()
-        # Call a dict and remove the if functions
-        if self.xunit == 'V':
-            df_['U_V'] = df['U']
+        The time axis can only be created when a (scan) rate is given in the plot, i.e., 50 mV /s.
+        """
+        df = self.svgplot.df.copy()
+        df = self._add_U_axis(df)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:135]:
```suggestion
        self._add_U_axis(df)
```

File: svgdigitizer/electrochemistry/cv.py [hunk 39]
```diff
@@ -19,110 +19,183 @@
...
 
-        if self.xunit == 'mV':
-            df_['U_V'] = df['U']/1E3
+        df = self._add_I_axis(df)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:137]:
```suggestion
        self._add_I_axis(df)
```

File: svgdigitizer/electrochemistry/cv.py [hunk 40]
```diff
@@ -19,110 +19,183 @@
...
 
-        df_['U_mV'] = df_['U_V']*1E3
+        df = self._add_time_axis(df)
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:139]:
```suggestion
        self._add_time_axis(df)
```

File: svgdigitizer/electrochemistry/cv.py [hunk 41]
```diff
@@ -19,110 +19,183 @@
...
-        if self.yunit == 'A':
-            df_['I_A'] = df['I']
+        q = 1 * CV.get_axis_unit(self.svgplot.units['x'])
+        # Convert the axis unit to SI unit V and use the value
+        # to convert the potential values in the df to V
+        df['U'] = df['x'] * q.to(u.V).value
 
-        if self.yunit == 'mA':
-            df_['I_A'] = df['I']/1E3
+        return df
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:153]:
```suggestion
```

File: svgdigitizer/electrochemistry/cv.py [hunk 42]
```diff
@@ -19,110 +19,183 @@
...
+            conversion_factor = q.to(u.A / u.m**2)
+        else:
+            conversion_factor = q.to(u.A)
 
-        df_['I_mA'] = df_['I_A']*1E3
-        df_['I_uA'] = df_['I_A']*1E6
+        df[self.axis_properties['y']['dimension']] = df['y'] * conversion_factor
 
-        return df_[['I_A', 'I_mA', 'I_uA']]
+        return df
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:169]:
```suggestion
```

File: svgdigitizer/electrochemistry/cv.py [hunk 43]
```diff
@@ -19,110 +19,183 @@
...
-        plt.ylabel(f'{self.ylabel} / {self.yunit}')
-
-    def create_csv(self, filename):
-        csvfile = Path(filename).with_suffix('.csv')
-        self.df.to_csv(csvfile, index=False)
+        df['deltaU'] = abs(df['U'].diff().fillna(0))
+        df['cumdeltaU'] = df['deltaU'].cumsum()
+        df['t'] = df['cumdeltaU'] / float(self.rate.to(u.V / u.s).value)
+
+        return df
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:179]:
```suggestion
```

Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:179]:
```suggestion
```

**Review by saraedum — APPROVED** [review]

File: svgdigitizer/electrochemistry/cv.py [hunk 44]
```diff
@@ -19,110 +19,179 @@
...
-        units given in the figure description.
+        Add a voltage column to the dataframe `df`, based on the :meth:`get_axis_unit` of the x axis.
         '''
-        df_ = df.copy()
-        if self.yunit == 'A':
-            df_['I_A'] = df['I']
+        q = 1 * CV.get_axis_unit(self.svgplot.units['x'])
+        # Convert the axis unit to SI unit V and use the value
+        # to convert the potential values in the df to V
+        df['U'] = df['x'] * q.to(u.V).value
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:152]:
```suggestion
```

File: svgdigitizer/electrochemistry/cv.py [hunk 45]
```diff
@@ -19,110 +19,179 @@
...
+        if 'm2' in str(q.unit):
+            conversion_factor = q.to(u.A / u.m**2)
+        else:
+            conversion_factor = q.to(u.A)
 
-        df_['I_mA'] = df_['I_A']*1E3
-        df_['I_uA'] = df_['I_A']*1E6
+        df[self.axis_properties['y']['dimension']] = df['y'] * conversion_factor
 
-        return df_[['I_A', 'I_mA', 'I_uA']]
```
Comment (saraedum) [inline svgdigitizer/electrochemistry/cv.py:168]:
```suggestion
```
